node/average.go

```diff
@@ -0,0 +1,111 @@
+package node
+
+import (
+	"context"
+
+	"github.com/pegnet/pegnetd/fat/fat2"
+)
+
+var AveragePeriod = uint64(288)         // Our Average Period is 2 days (144 10 minute blocks per day)
+var AverageRequired = AveragePeriod / 2 // If we have at least half the rates, we can do conversions
+
+// getPegNetRateAverages
+// Gets all the rates for the AveragePeriod (the number of blocks contributing to the average), and computes
+// the average rate for all assets.  If values are missing for an asset for any of the blocks in the AveragePeriod,
+// then don't allow conversions by setting the averate to zero for that asset
+//
+// Return a map of averages of the form map[fat2.PTicker]uint64
+//
+// Also note that if asked twice about the same height, we cache the response.
+func (d *Pegnetd) GetPegNetRateAverages(ctx context.Context, height uint32) (Avg interface{}) {
+
+	if d.LastAveragesHeight == height { //                      If a cache hit is detected, return the cache value
+		return d.LastAverages
+	}
+
+	ratesOverPeriod := d.LastAveragesData //                    First collect all the values over the blocks
+	averages := map[fat2.PTicker]uint64{} //                      in the average period, then compute the averages
+	if ratesOverPeriod == nil {           //                    If no map exists yet
+		ratesOverPeriod = map[fat2.PTicker][]uint64{} //          create one.
+	}
+
+	defer func() { //                                           Always set up the cache when exiting the routine
+		d.LastAveragesData = ratesOverPeriod //                   Save the data we used to create averages
+		d.LastAveragesHeight = height        //                   Save the height of this data
+		d.LastAverages = averages            //                   Save the averages we computed
+	}()
+
+	// collectRatesAtHeight
+	// This routine collects all the data used to compute an average.  If any data is missing, then
+	// that data is represented by a zero.
+	collectRatesAtHeight := func(h uint32) {
+		for k := range ratesOverPeriod { //                   Make sure there is room for a new height
+			for len(ratesOverPeriod[k]) >= int(AveragePeriod) { //  If at the limit or above,
+				copy(ratesOverPeriod[k], ratesOverPeriod[k][1:])                    // Shift data down 1 element
+				ratesOverPeriod[k] = ratesOverPeriod[k][:len(ratesOverPeriod[k])-1] //   And drop off the last value
+			}
+		}
+
+		if rates, err := d.Pegnet.SelectRates(ctx, h); err != nil { // Pull the rates out of the database at each
+			panic("no recovery from a database error getting rates")
```

Comment:
This will break/terminate the pegnetd app with error.
Is it better to return the error than panic termination here?

Reply:
`SelectRates()` will return just empty slices if there are no rows at specific height.
It is totally good to panic the context if db is corrupted.